src/main/java/org/wildfly/security/authz/jacc/ElytronPolicyConfiguration.java

```diff
@@ -0,0 +1,231 @@
+
+        this.excludedPermissions.add(permission);
+    }
+
+    @Override
+    public void addToExcludedPolicy(PermissionCollection permissions) throws PolicyContextException {
+        checkNotNullParam("permissions", permissions);
+
+        Enumeration<Permission> elements = permissions.elements();
+
+        while (elements.hasMoreElements()) {
+            addToExcludedPolicy(elements.nextElement());
+        }
+    }
+
+    @Override
+    public void addToRole(String roleName, Permission permission) throws PolicyContextException {
+        checkNotNullParam("roleName", roleName);
+        checkNotNullParam("permission", permission);
+        checkIfInOpenState();
+
+        this.rolePermissions.computeIfAbsent(roleName, s -> new Permissions()).add(permission);
+    }
+
+    @Override
+    public void addToRole(String roleName, PermissionCollection permissions) throws PolicyContextException {
+        checkNotNullParam("roleName", roleName);
+        checkNotNullParam("permissions", permissions);
+
+        Enumeration<Permission> elements = permissions.elements();
+
+        while (elements.hasMoreElements()) {
+            addToRole(roleName, elements.nextElement());
+        }
+    }
+
+    @Override
+    public void addToUncheckedPolicy(Permission permission) throws PolicyContextException {
+        checkNotNullParam("permission", permission);
+        checkIfInOpenState();
+
+        this.uncheckedPermissions.add(permission);
+    }
+
+    @Override
+    public void addToUncheckedPolicy(PermissionCollection permissions) throws PolicyContextException {
+        checkNotNullParam("permissions", permissions);
+
+        Enumeration<Permission> elements = permissions.elements();
+
+        while (elements.hasMoreElements()) {
+            addToUncheckedPolicy(elements.nextElement());
+        }
+    }
+
+    @Override
+    public void commit() throws PolicyContextException {
+        synchronized (this) {
+            if (isDeleted()) {
+                throw log.authzInvalidStateForOperation(this.state.name());
+            }
+
+            transitionTo(State.IN_SERVICE);
+        }
+    }
+
+    @Override
+    public void delete() throws PolicyContextException {
+        synchronized (this) {
+            transitionTo(State.DELETED);
+            this.uncheckedPermissions = new Permissions();
+            this.excludedPermissions = new Permissions();
+            this.rolePermissions.clear();
+            this.linkedPolicies.remove(this);
+        }
+    }
+
+    @Override
+    public String getContextID() throws PolicyContextException {
+        return this.contextId;
+    }
+
+    @Override
+    public boolean inService() {
+        return State.IN_SERVICE.equals(this.state);
+    }
+
+    @Override
+    public void linkConfiguration(PolicyConfiguration link) throws PolicyContextException {
+        checkNotNullParam("link", link);
+        checkIfInOpenState();
+
+        synchronized (this.linkedPolicies) {
```

Comment:
This however is not fine.  You're synchronizing on the collection that the field points to, but you're updating that same field inside the block.  Thus anyone reading the field may get an outdated copy.